apollo-ast/src/main/kotlin/com/apollographql/apollo3/ast/transformation/add_defer_labels.kt

```diff
@@ -0,0 +1,57 @@
+package com.apollographql.apollo3.ast.transformation
+
+import com.apollographql.apollo3.ast.GQLArgument
+import com.apollographql.apollo3.ast.GQLArguments
+import com.apollographql.apollo3.ast.GQLDirective
+import com.apollographql.apollo3.ast.GQLFragmentDefinition
+import com.apollographql.apollo3.ast.GQLNode
+import com.apollographql.apollo3.ast.GQLOperationDefinition
+import com.apollographql.apollo3.ast.GQLStringValue
+import com.apollographql.apollo3.ast.NodeTransformer
+import com.apollographql.apollo3.ast.TransformResult
+import com.apollographql.apollo3.ast.transform
+
+fun addDeferLabel(operation: GQLOperationDefinition): GQLOperationDefinition {
```

Comment:
Why this is required? From the [RFC](https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md), I understand the label is optional. If that's the case, I'd avoid modifying the user document too much. We do it for `__typename` and some ids but the less we modify the better.

From my very superficial understanding of the RFC, the label is only required when 2 deferred fragments have the same path? If that's the case, I'd rather much have a validation rule to fail compilation if 2 fragments with the same paths don't have a "label" (and I guess one having a label but not the other is enough to disambiguate)

Reply:
Correct, it is optional. The [idea](https://github.com/apollographql/apollo-kotlin/pull/3858/files#diff-3fe0960286149c9e1b604842ea5bd3e18a326f6f59fac3cce97e3ad53e0714e8R373) is to use the label as a way to know when to / not to parse a fragment's fields. Since labels must be unique across a document they are a good match for this - from the RFC they're used "to identify the data from responses and associate it with the corresponding defer directive".

Indeed, to avoid automatically adding labels, we may be able to use "path+label" instead, and ensure that all "path+label" are unique. Inside the generated adapter code, I'm not yet sure if it's easy to know the current path?

However, going this route, whenever several defer fragments are present at the same level, the path will be the same, and it will be up to the user to resolve the ambiguity by adding a unique label. From a user's perspective, wouldn't it be nicer for the SDK to take care of this automatically?

Reply:
> From a user's perspective, wouldn't it be nicer for the SDK to take care of this automatically?

I'd prefer to give the user an error and let them come up with the label. This way, they can copy/paste the GraphQL document to/from GraphiQL/Studio and it'll work out of the box. We can add "auto-label" down the road if it comes up a lot but if we start with "auto-label" from day 1, we will never be able to go back if we ever regret the decision.

Reply:
I see what you mean. Just to be 100% thorough, OTOH, there could be cases where a valid query (that works in GraphiQL/Studio) would be rejected by the compiler, as the rule to have unique path+label is not in the spec but specific to us.

Reply:
>  the rule to have unique path+label is not in the spec but specific to us.

Would it make sense to put it in the spec? 

Something like below sounds ambiguous:

```graphql
query GetHero {
  hero {
    ...heroDetails1 @defer
    ...heroDetails2 @defer
  }
}
```

Reply:
A follow-up: this was not 100% clear to me before but now I realize in any case, we *must* use path+label because whenever arrays are involved, the label alone is not enough. In [this example](https://github.com/apollographql/apollo-kotlin/pull/3858/files#diff-3fe0960286149c9e1b604842ea5bd3e18a326f6f59fac3cce97e3ad53e0714e8R351), the fields for the `ComputerFields` fragment must be parsed for `["computers",0]` and not for `["computers",1]`.

Reply:
Update: removed the automatically computed label (and created #3938).